contract/models/contract.py

```diff
@@ -181,68 +214,62 @@ def _prepare_invoice_line(self, line, invoice_id):
         })
         return invoice_line_vals
 
-    @api.model
-    def _prepare_invoice(self, contract):
-        if not contract.partner_id:
+    @api.multi
+    def _prepare_invoice(self):
+        self.ensure_one()
+        if not self.partner_id:
             raise ValidationError(
                 _("You must first select a Customer for Contract %s!") %
-                contract.name)
-        journal = contract.journal_id or self.env['account.journal'].search(
+                self.name)
+        journal = self.journal_id or self.env['account.journal'].search(
             [('type', '=', 'sale'),
-             ('company_id', '=', contract.company_id.id)],
+             ('company_id', '=', self.company_id.id)],
             limit=1)
         if not journal:
             raise ValidationError(
                 _("Please define a sale journal for the company '%s'.") %
-                (contract.company_id.name or '',))
+                (self.company_id.name or '',))
         currency = (
-            contract.pricelist_id.currency_id or
-            contract.partner_id.property_product_pricelist.currency_id or
-            contract.company_id.currency_id
+            self.pricelist_id.currency_id or
+            self.partner_id.property_product_pricelist.currency_id or
+            self.company_id.currency_id
         )
         invoice = self.env['account.invoice'].new({
-            'reference': contract.code,
+            'reference': self.code,
             'type': 'out_invoice',
-            'partner_id': contract.partner_id.address_get(
+            'partner_id': self.partner_id.address_get(
                 ['invoice'])['invoice'],
             'currency_id': currency.id,
             'journal_id': journal.id,
-            'date_invoice': contract.recurring_next_date,
-            'origin': contract.name,
-            'company_id': contract.company_id.id,
-            'contract_id': contract.id,
+            'date_invoice': self.recurring_next_date,
+            'origin': self.name,
+            'company_id': self.company_id.id,
+            'contract_id': self.id,
+            'user_id': self.partner_id.user_id.id,
         })
         # Get other invoice values from partner onchange
         invoice._onchange_partner_id()
-        return invoice._convert_to_write(invoice._cache)
+        return invoice._convert_to_write(
+            {name: invoice[name] for name in invoice._cache})
```

Comment:
Because in the case of empty values for many2one records, the cache contains an empty tuple instead of an empty recordset. The invoice record contains the correct value.

Before I added this change, this line broke the test. I took the modification from the Odoo core addons where it is used in several places.